src/Language/PureScript/Renamer.hs

```diff
@@ -109,30 +110,36 @@ renameInModules :: [Module Ann] -> [Module Ann]
 renameInModules = map go
   where
   go :: Module Ann -> Module Ann
-  go m@(Module _ _ _ _ _ _ _ _ decls) = m { moduleDecls = map (renameInDecl' (findDeclIdents decls)) decls }
-
-  renameInDecl' :: [Ident] -> Bind Ann -> Bind Ann
-  renameInDecl' scope = runRename scope . renameInDecl True
+  go m@(Module _ _ _ _ _ exports _ foreigns decls) = m { moduleExports, moduleDecls }
+    where
+    (moduleDecls, moduleExports) = runRename (findDeclIdents decls ++ foreigns) $
+      (,) <$> renameInDecls True decls <*> traverse lookupIdent exports
 
 -- |
--- Renames within a declaration. isTopLevel is used to determine whether the
--- declaration is a module member or appearing within a Let. At the top level
--- declarations are not renamed or added to the scope (they should already have
--- been added), whereas in a Let declarations are renamed if their name shadows
--- another in the current scope.
+-- Renames within a list of declarations. isTopLevel is used to determine
+-- whether the declarations are module members or appearing within a Let. At
+-- the top level declarations are only renamed or added to the scope if they
+-- are named with GenIdents (otherwise they should already have been added),
+-- whereas in a Let declarations are renamed if their name shadows another in
+-- the current scope. In order to prevent inner bindings earlier in the list
+-- from shadowing bindings later in the list, first all of the declarations
+-- are considered for renaming, and then all of the values are recursed into.
```

Comment:
> In order to prevent inner bindings earlier in the list from shadowing bindings later in the list, first all of the declarations are considered for renaming, and then all of the values are recursed into.

Could you clarify this point further because I'm not familiar with the code outside of this file?

Is this in case something like this occurs?
```purescript
foo :: String -> String
foo a =
  -- does it use top-level `a` 
  -- or foo's `a`? 
  a <> "something"

a :: String
a = "text value

Reply:
Yes, that's exactly what I'm talking about there. The `a` in `foo` is an inner binding that shadows the top-level `a`. The renamer should rename the `a` in `foo` to `a1`. I think the CoreImp optimizer expects the code it receives to have no shadowing, so it's not just a style thing (even though in JavaScript shadowing works just fine).

Reply:
Gotcha. Thanks for clarifying!